core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -200,14 +229,32 @@ class LogSegment(val log: FileMessageSet,
     val truncated = log.sizeInBytes - validBytes
     log.truncateTo(validBytes)
     index.trimToValidSize()
+    // A normally closed segment always appends the biggest timestamp ever seen into log segment, we do this as well.
+    timeIndex.maybeAppend(maxTimestampSoFar, offsetOfMaxTimestamp)
+    timeIndex.trimToValidSize()
     truncated
   }
 
+  def loadTimestamps(readToLogEnd: Boolean = false) {
+    // Get the last time index entry. If the time index is empty, it will return (-1, baseOffset)
+    if (log.iterator.hasNext)
+      rollingBaseTimestamp = log.iterator.next.message.timestamp
```

Comment:
maybeRoll() is only called on the active segment. So, if we get first message on demand when maybeRoll() is called, we can avoid reading the first message on all old segments.